### PR TITLE
Release Notes: Render excerpt in list with reduced formatting, show read more. 

### PR DESCRIPTION
**Pull request by nkuehn**
This partially supports #360 but not completely. 

Because it introduces the infrastructure for the same use case this also addresses #321  (proper rendering of markdown in API RAML description texts)

Rendering excerpts turned out to not be possible using gatsby's original `excerpt_separator` configuration because this setting is only applicable to the remark plugin, but not to MDX.  The MDX plugin does only support plaintext excerpts that are chopped off at a specific number of characters. 

Also, I originally intended to only allow rendering a subset of markdown in the release notes list view.  We left the requirement behind to save complexity, but now it turned out it's actually easier to do it this way because we have to build a "safe" limited markdown renderer anyways for other places where small fragments of MD are placed into other layouts. 

Solution here is to "chop" the raw MDX into halves ourselves and providing the first part as a new property on GraphQL.  Plus a boolean convenience flag that shows that there is not more in the full body (it's hard to compare because the full body is javascript source code already). 

The Component that renders a release note now behaves differently depending on whether the full body was passed or just an excerpt.  

If only the excerpt is passed it's rendering this using a plain markdown renderer that has all HTML / MD constructs "disabled" that could lead to problems e.g. if the content would be placed into a table cell or inside other blocks.

### CODE REVIEW

File: packages/gatsby-theme-docs/src/layouts/internals/layout-release-note.js [hunk 1]
```diff
@@ -59,7 +66,9 @@ ReleaseNote.propTypes = {
   description: PropTypes.string.isRequired,
   type: PropTypes.oneOf(['feature', 'enhancement', 'fix']).isRequired,
   topics: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
-  body: PropTypes.string.isRequired,
+  body: PropTypes.string,
+  rawExcerpt: PropTypes.string,
```
Comment (nkuehn) [inline packages/gatsby-theme-docs/src/layouts/internals/layout-release-note.js:70]: This is not ideal, but since the GraphQL data source is already strongly typed I think it's acceptable.  We could write a custom validation function to make sure either one is passed if we think this component will be re-used a lot.

File: packages/gatsby-theme-docs/src/templates/release-notes-list.js [hunk 2]
```diff
@@ -73,7 +62,8 @@ ReleaseNotesListTemplate.propTypes = {
           description: PropTypes.string.isRequired,
           type: PropTypes.string.isRequired,
           topics: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
-          body: PropTypes.string.isRequired,
+          rawExcerpt: PropTypes.string.isRequired,
```
Comment (nkuehn) [inline packages/gatsby-theme-docs/src/templates/release-notes-list.js:65]: Side effect: not loading the full content is good for the page weight too.

**Comment by nkuehn** [comment]
Side remark: We shouldn't forget to introduce visual regression tests for the release notes and consider expanding to some of the API documentation cases again. 
I expected this to break the VRTs for the API docs but it doesn't  - the cases it changes aren't tested.

File: websites/docs-smoke-test/src/releases/excerpt-rendering-test.mdx [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+  <dd>Definition of the Term</dd>
+</dl>
+
+separator line below this
+
+----
+
+<!--more-->
+
+----
```
Comment (emmenko) [inline websites/docs-smoke-test/src/releases/excerpt-rendering-test.mdx:49]: So is this format how you want people to use the "write more"?

Feels a bit fragile, maybe it's just my impression.

File: packages/gatsby-theme-docs/src/layouts/internals/layout-release-note.js (hunk 1 above)
Comment (emmenko) [reply packages/gatsby-theme-docs/src/layouts/internals/layout-release-note.js:70]: Hmm but both are required fields, aren't they?

I would expect the logic to be (pseudo-code)

```
if hasMore
  render excerpt
  render read more
else
  render body
```

File: packages/gatsby-theme-docs/src/layouts/internals/layout-release-note.js [hunk 4]
```diff
@@ -59,7 +66,9 @@ ReleaseNote.propTypes = {
   description: PropTypes.string.isRequired,
   type: PropTypes.oneOf(['feature', 'enhancement', 'fix']).isRequired,
   topics: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
-  body: PropTypes.string.isRequired,
+  body: PropTypes.string,
```
Comment (davifantasia) [inline packages/gatsby-theme-docs/src/layouts/internals/layout-release-note.js:69]: Since you're no longer querying for `body` from `src/templates/release-notes-list.js`, this prop is no longer necessary.

File: packages/gatsby-theme-docs/src/layouts/internals/layout-release-note.js (hunk 1 above)
Comment (nkuehn) [reply packages/gatsby-theme-docs/src/layouts/internals/layout-release-note.js:70]: The excerpt is going through a different renderer now than the full body, so in the list view it's always render excerpt and the "If" just covers whether to show the link or not.  

If the body is made required, we would pump up the data size of the page query of the list view with all the unused bodies.

File: websites/docs-smoke-test/src/releases/excerpt-rendering-test.mdx (hunk 3 above)
Comment (nkuehn) [reply websites/docs-smoke-test/src/releases/excerpt-rendering-test.mdx:49]: We went through that decisionmaking in a PR that was explicitly about agreeing on the format specification. 

That format is the current format (but that could be migrated).   One requirement is that it must be handled opaque by Markdown preview implementations, so it must be either a valid but invisible markdown element or an HTML comment.  

I proposed a different format, too (e.g. just a horizontal line) but the tech writers preferred this format becuause it's verbally explicit.

Comment (emmenko) [reply websites/docs-smoke-test/src/releases/excerpt-rendering-test.mdx:49]: Thanks for the heads up.

File: packages/gatsby-theme-docs/src/layouts/internals/layout-release-note.js (hunk 4 above)
Comment (nkuehn) [reply packages/gatsby-theme-docs/src/layouts/internals/layout-release-note.js:69]: The component is used on the details page, too so it needs to be able to accept either a full body or just an excerpt. 

But I could refactor it in a way that the excerpt is passed as the body and in addition we pass just the boolean that tells which renderer to use.

Comment (emmenko) [reply packages/gatsby-theme-docs/src/layouts/internals/layout-release-note.js:69]: > But I could refactor it in a way that the excerpt is passed as the body and in addition we pass just the boolean that tells which renderer to use.

Yes that would be preferrable. This component should not concern itself where it's rendered, the parent knows the context.

Comment (nkuehn) [reply packages/gatsby-theme-docs/src/layouts/internals/layout-release-note.js:69]: @davifantasia  let's talk about this one in our half hour today - I'm confused about my own implementation now, maybe it has to do with work you did on the details pages. 

There isn't any second use of this component any more in the codebase.

File: packages/gatsby-theme-docs/src/layouts/internals/layout-release-note.js [hunk 5]
```diff
@@ -60,6 +67,8 @@ ReleaseNote.propTypes = {
   type: PropTypes.oneOf(['feature', 'enhancement', 'fix']).isRequired,
   topics: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
   body: PropTypes.string.isRequired,
+  bodyRenderer: PropTypes.oneOf(['mdx', 'excerpt']).isRequired,
```
Comment (emmenko) [inline packages/gatsby-theme-docs/src/layouts/internals/layout-release-note.js:70]: Hmm, let me push a commit...

Comment (emmenko) [reply packages/gatsby-theme-docs/src/layouts/internals/layout-release-note.js]: So: https://github.com/commercetools/commercetools-docs-kit/pull/375/commits/3d32a5669530177268f50ef765a9e93870fde48a

TL;DR: The parent should just format the body and pass it as a children.

File: packages/gatsby-theme-docs/src/templates/release-notes-list.js [hunk 6]
```diff
@@ -31,11 +32,13 @@ const ReleaseNotesListTemplate = (props) => (
...
-                  key={releaseNote.slug}
-                  bodyRenderer="excerpt"
-                  {...releaseNote}
-                />
+                <LayoutReleaseNote key={releaseNote.slug} {...releaseNote}>
+                  <Markdown.TypographyPage>
+                    <section>
+                      {markdownFragmentToReact(releaseNote.body)}
+                    </section>
+                  </Markdown.TypographyPage>
```
Comment (emmenko) [inline packages/gatsby-theme-docs/src/templates/release-notes-list.js:40]: The parent can decide how the render the body and pass it as a child.

File: packages/gatsby-theme-docs/src/templates/release-notes-list.js [hunk 7]
```diff
@@ -31,11 +32,13 @@ const ReleaseNotesListTemplate = (props) => (
...
               props.data.allReleaseNotePage.nodes &&
               props.data.allReleaseNotePage.nodes.map((releaseNote) => (
-                <LayoutReleaseNote
-                  key={releaseNote.slug}
-                  bodyRenderer="excerpt"
-                  {...releaseNote}
-                />
+                <LayoutReleaseNote key={releaseNote.slug} {...releaseNote}>
+                  <Markdown.TypographyPage>
+                    <section>
```
Comment (emmenko) [inline packages/gatsby-theme-docs/src/templates/release-notes-list.js:37]: @nkuehn should these elements have some more contextual info? Like ids, labels, ...

File: packages/gatsby-theme-docs/src/layouts/internals/layout-release-note.js (hunk 4 above)
Comment (emmenko) [reply packages/gatsby-theme-docs/src/layouts/internals/layout-release-note.js]: See https://github.com/commercetools/commercetools-docs-kit/pull/375#discussion_r411267318

File: packages/gatsby-theme-docs/src/layouts/internals/layout-release-note.js [hunk 8]
```diff
@@ -45,24 +45,23 @@ const ReleaseNote = (props) => {
           ))}
         </Topics>
       )}
-
-      <div>
-        <MDXRenderer>{props.body}</MDXRenderer>
-      </div>
+      {props.children}
+      {props.hasMore ? <Link href={`${props.slug}`}>Read more...</Link> : null}
```
Comment (emmenko) [inline packages/gatsby-theme-docs/src/layouts/internals/layout-release-note.js:49]:
```suggestion
      {props.hasMore ? <Link href={props.slug}>Read more...</Link> : null}
```
